Feature: Share server URL by QR

shaunburch opened the pull request:
<!--
- Add the issue number here.

- If you haven't solved the issue completely use "Linked issue #{issue_number}.
- After solving the issue completely change it to "Fixes #{issue_number}.
-->
Fixes #1553 

<!-- Add here what changes were made in this issue and if possible provide links. -->
- Added: [ZXing](https://github.com/zxing/zxing) for QR generation
- Added: [GenerateQR](core/src/main/java/org/kiwix/kiwixmobile/core/qr/GenerateQR.kt) to generate QR code
- Added: QR shown on "WiFi Hotspot" (`ZimHostFragment`) screen when server started 
- Added: QR shown on "Kiwix Hotspot" (`HotspotNotificationManager`) notification
<!-- If possible, please add relevant screenshots / GIFs -->

**Screenshots** 
| Small Phone, 360x640dp, API 28 | Pixel 6, 412x915dp, API 34) |
|--|--|
![small-phone-wifi-hotspot-qr](https://github.com/kiwix/kiwix-android/assets/45771676/ea896188-1377-41db-b720-9f1c9425dcee) | ![Kiwix-LargePhone](https://github.com/kiwix/kiwix-android/assets/45771676/bf6a2217-752e-40cb-bfdf-6f5465c41d54)

### Video
[kiwix-demo.webm](https://github.com/kiwix/kiwix-android/assets/45771676/4ba5dd6f-c259-4646-b117-b4cc6312ac84)

shaunburch commented:
@kelson42 PR for review to add QR for connecting.

kelson42 commented:
@shaunburch thank you very much for your PR. At this stage I worry about so many new dependences, in particular zxing which seems already almost deprecated. Can we not do simpler to make a substainable PR?

shaunburch commented:
> @shaunburch thank you very much for your PR. At this stage I worry about so many new dependences, in particular zxing which seems already almost deprecated. Can we not do simpler to make a substainable PR?

@kelson42 I'm not sure I follow. I believe ZXing is the only dependency added. Is there another QR generation lib you'd prefer?

kelson42 commented:
> > @shaunburch thank you very much for your PR. At this stage I worry about so many new dependences, in particular zxing which seems already almost deprecated. Can we not do simpler to make a substainable PR?
> 
> @kelson42 I'm not sure I follow. I believe ZXing is the only dependency added. Is there another QR generation lib you'd prefer?

The best would be to use something native of course, but if we really need an external one, the please one which is actively maintained and which support Android14.

shaunburch commented:
@kelson42 Still not sure I understand this feedback:

> The best would be to use something native of course

Native as in C? Or Java? ZXing is Java.
<img width="318" alt="Screenshot 2024-06-10 at 8 55 56 AM" src="https://github.com/kiwix/kiwix-android/assets/45771676/6e7b3be8-59d0-43e1-bc1d-61de828bfd3f">


> if we really need an external one, the please one which is actively maintained

I'm not aware of an internal dependency for QR generation. ZXing is maintained by Google (latest release Jan 30), and the most widely used libraries to encode/decode QR (+ other barcode formats) for Android apps.
<img width="745" alt="Screenshot 2024-06-10 at 8 57 40 AM" src="https://github.com/kiwix/kiwix-android/assets/45771676/808af61d-c4c4-4b60-8edb-b9e9c878ab3a">

There is the [Google Vision / ML Kit API](https://developers.google.com/ml-kit/vision/text-recognition/v2/android) but that requires Play Services and I saw the note in README.md about variant created over use of Play Services; opted to take the path more available to users.


>  and which support Android14.
<img width="420" alt="API 34 - Android 14" src="https://github.com/kiwix/kiwix-android/assets/45771676/3c6c63d3-9db4-4750-8c23-12f00345b57d">

Development was done on Android 14 (API 34)


If you have a specific implementation for QR in mind, I'll gladly refactor this to use it – but I'm otherwise confused by this feedback.

MohitMaliFtechiz commented:
@kelson42 I have checked for other dependencies of the QR code generator, but they are not maintained and also they are using the `ZXing` under the hood. `ZXing` is the only lib that is updated this year. For the Android 14 supports: `Zxing` supports the generating of QR code in Android 14 as shown in the below image. They mentioned in their README file that the [scanning of QR code will not support for Android 14](https://github.com/zxing/zxing?tab=readme-ov-file#project-in-maintenance-mode-only-barcode-scanner-not-available).

![1718018119000](https://github.com/kiwix/kiwix-android/assets/34593983/e50319d0-6e09-4401-af93-bfcdfce5cfdf)

There are no default classes for QR codes in Android, if we go for the native approach we have to handle the `error correction`, `Encoding Modes`, etc like it is handled in the `ZXing` lib (Meaning we have to make the same code like `QRCodeWriter` in the `Zxing` lib).

Apart from the Zxing, there is only one library I have found that is updated last month [OkapiBarcode](https://github.com/woo-j/OkapiBarcode), but it has a linear barcode format.

![Screenshot from 2024-06-10 18-54-01](https://github.com/kiwix/kiwix-android/assets/34593983/4a508b06-82ee-42d5-a7f6-f8ebf6545ff8)

MohitMaliFtechiz commented:
> Native as in C? Or Java? ZXing is Java.

@shaunburch Here native means Android itself.

MohitMaliFtechiz commented:
> There is the [Google Vision / ML Kit API](https://developers.google.com/ml-kit/vision/text-recognition/v2/android) but that requires Play Services

@shaunburch AFAIK Google ML kit has the [barcode scanning](https://developers.google.com/ml-kit/vision/barcode-scanning) feature not the barcode generator please see https://developers.google.com/ml-kit.

shaunburch commented:
> @shaunburch AFAIK Google ML kit has the [barcode scanning](https://developers.google.com/ml-kit/vision/barcode-scanning) feature not the barcode generator please see https://developers.google.com/ml-kit.

That appears correct, recognition / decode only.

MohitMaliFtechiz commented:
> I'm not aware of an internal dependency for QR generation. ZXing is maintained by Google (latest release Jan 30), and the most widely used libraries to encode/decode QR (+ other barcode formats) for Android apps.

@shaunburch IMO forking is not a suitable solution for this feature, here we only generate the QR code for the URL which is a very simple QR code, and generating this type of simple QR code does not impact due to the Android versions(Like for Android 14). Also, the library is maintained by Google and will not add support for scanning the barcode since they introduced it in the Android ML kit(This is the only reason they have not added support for Android 14). But the QR code generator will work at least for now(Since there is no library in the ML kit for generating the barcode, and they do not add it in the ML kit because this library already exists, and it is the only library that is core for generating the barcode since other libraries are using this under the hood). So IMO we should use this library @kelson42 What do you think?

kelson42 commented:
Then we are good regarding the dependency, I guess. I will Look to the Feature Inseln.

MohitMaliFtechiz commented on app/src/androidTest/java/org/kiwix/kiwixmobile/webserver/ZimHostFragmentTest.kt:
```diff
@@ -160,6 +160,10 @@ class ZimHostFragmentTest {
 
+        // Check Share by QR code
+        tapShareByQr()
+        assertQRDialogShown()
```
The test case is failing due to the QR code showing on the screen, please add the `pressBack()` after `assertQRDialogShown()` so that the QR code will remove from the screen and the test case will pass.

MohitMaliFtechiz commented on app/src/main/java/org/kiwix/kiwixmobile/qr/ShareByQRCodeDialog.kt:
```diff
@@ -0,0 +1,60 @@
+   * Load the QR code for the given [uri].
+   */
+  private fun loadQrCode(uri: String) {
+    val qr = GenerateQR().createQR(uri)
```
Please use Injection instead of directly creating the objects.

MohitMaliFtechiz commented on core/src/main/res/layout/dialog_share_by_qr_code.xml:
```diff
@@ -0,0 +1,54 @@
+    app:layout_constraintEnd_toEndOf="parent"
+    app:layout_constraintStart_toStartOf="parent"
+    app:layout_constraintTop_toBottomOf="@id/qr_code"
+    tools:text="Scan the QR Above" />
```
Do not hardcode any string.

shaunburch replied:
I will remove it, but want to confirm, this applies to `tools:text` strings used only in preview?

MohitMaliFtechiz commented on app/src/main/java/org/kiwix/kiwixmobile/qr/ShareByQRCodeDialog.kt:
```diff
@@ -0,0 +1,60 @@
+import org.kiwix.kiwixmobile.core.databinding.DialogShareByQrCodeBinding
+import org.kiwix.kiwixmobile.core.qr.GenerateQR
+
+class ShareByQRCodeDialog : DialogFragment() {
```
Please move this class to the `core` module like the `GenerateQR` class. Also, rename this class to `QRCodeDialog` since this only shows the QR code and does not share anything, so `QRCodeDialog` would be better for this.

shaunburch replied:
@MohitMaliFtechiz is there documentation on the concerns for each module? 

I noticed `core` seemed to center around domain logic but also had Android views. Opted here to keep Android dependent code out of domain.

Happy to move, just hoping to understand intent of modules better.

MohitMaliFtechiz replied:
@shaunburch `core` module is the base for both app and custom apps. We put the common functionality in the `core` module to use it in both modules that's why you are seeing some Android views in that module. For the server functionality, we will enable this feature in custom apps that's why I am asking you to put this in the core module.

MohitMaliFtechiz reviewed (changes requested):
@shaunburch Thank you for your work. Just need small changes and then we are good to go.

kelson42 commented:
@shaunburch @MohitMaliFtechiz Finally here is my feedback regarding UX. I propose to change current behaviour to:
* Remove the QR button (beside the "Share" button)
* Just always display the QR code below the text "server http://...." (and above the list of ZIM files)
* Display the QR code as well in the notification

Does that make sense?

MohitMaliFtechiz commented:
Sounds good to me, @shaunburch can you please implement this and test with the small screen size devices that QR code is easily scannable and the list of ZIM files will not go too down.

shaunburch commented:
> Display the QR code as well in the notification

@kelson42 This notification?

<img width="327" alt="hotspot-running" src="https://github.com/kiwix/kiwix-android/assets/45771676/07602da9-7007-4ab9-93f3-4f6631dffe42">

shaunburch commented:
@kelson42 @MohitMaliFtechiz Placing QR below server text, using [QR Code Resolution Requirements](https://www.sproutqr.com/blog/qr-code-minimum-size#toc-qr-code-resolution-requirements) of `76dp` x `76dp`, works well enough in portrait but eats into screen real estate for files in landscape. Thoughts?
<img width="400" alt="76dp-landscape" src="https://github.com/kiwix/kiwix-android/assets/45771676/a2879426-5b95-4bfd-95ba-02046b4f8869">
_Small Phone (360 x 640 dp, xhdpi)_

kelson42 commented:
@shaunburch Thenspace between the text and the button could not have a vertical scrollbar? The qr code should also be bigger, why it is so small?

shaunburch commented:
> Thenspace between the text and the button could not have a vertical scrollbar?

Placing a recycle view (ZIM files) within a scroll view that encompasses it and QR creates a nested vertical scroll (two things compete for handling scroll gesture). Scrolling ZIM files would only scroll those. Scrolling QR would scroll it and entirety of ZIM files. Alternatively the QR could be made a list item in the recyclerview but would scroll with ZIM files and not always be shown (affixed to start of list). 

> The qr code should also be bigger, why it is so small?

It is small to illustrate even at minimum res (76x76) the real estate used in landscape / small devices.

kelson42 commented:
@MohitMaliFtechiz This becomes too technical to me, can you handle it (make a proposal) so the  QR code is nicely displayed just displayed on the top of the ZIM list?
@shaunburch What about the notification?

shaunburch commented:
@MohitMaliFtechiz Pushed commits to fix UI test (`pressBack()`) & other feedback.

@kelson42 Added QR to persistent Notification when hotspot activated:
| QR | Scanning QR |
|-|-|
| <img src="https://github.com/kiwix/kiwix-android/assets/45771676/6cd6b802-eda2-42bd-800f-950d07e2260e" width=300 />) | <img src="https://github.com/kiwix/kiwix-android/assets/45771676/f5fbd2f9-bc38-4303-9260-07e40cfe87f1" width=300 /> |

### Video
[hotspot-notification-with-qr.webm](https://github.com/kiwix/kiwix-android/assets/45771676/a089d2e3-a6ab-4cba-b0e2-9951dd304f25)

MohitMaliFtechiz commented on core/build.gradle.kts:
```diff
@@ -84,4 +84,7 @@ dependencies {
   testImplementation(Libs.kotlinx_coroutines_test)
   implementation(Libs.kotlinx_coroutines_android)
+
+  // ZXing
```
Please remove this comment as the `implementation(Libs.zxing)` already shows this is a Zxing library.

shaunburch replied:
🚮 Removed

shaunburch commented:
## Changes
- Removed `// Zxing` comment
- Created nested scrolling with QR above ZIM file `RecyclerView`
- Resized QR
- Using getSocketAddress to include port for persistent notification
- Removed unused code
- Updated test to verify QR _not_ shown on server stop (Including `SHORT_WAIT = 3_000L // 3s` as default waited 10s to verify)

## Video
| Small Phone | Pixel 6 |
|--|--|
<video src="https://github.com/kiwix/kiwix-android/assets/45771676/6d8004ff-6bbb-40da-94f0-4122a842deea" /> | <video src="https://github.com/kiwix/kiwix-android/assets/45771676/d99c84ff-4639-4272-a8cc-c5f685c04729" />

## Concerns
Nested scrolling creates competition between which view handles the scroll event and makes for an unintuitive UX for users.

MohitMaliFtechiz commented:
@shaunburch Thanks for the changes. Codewise now it looks good. The current UX helps to always show the QR code on the screen in portrait mode, and in landscape mode, the QR code will scroll up so that the zim file list will show. @kelson42 your opinion on the UX?

shaunburch commented:
> Edited: @shaunburch Please update the PR description, current description is outdated.

@MohitMaliFtechiz Descriptions updated 🙏

kelson42 commented:
We need to wait a bit (that we are over with patch releases) before merging this PR
